### ISSUE TITLE
trailing_closure warning not reported when using Xcode <= 12.4

### ISSUE THREAD

**Issue by revolter** [realm/SwiftLint#3592]
### New Issue Checklist

- [x] Updated SwiftLint to the latest version
- [x] I searched for [existing GitHub issues](https://github.com/realm/SwiftLint/issues)

### Describe the bug

A `trailing_closure` warning is not reported when using Xcode <= 12.4

##### Complete output when running SwiftLint, including the stack trace and command used

```bash
$ ./swiftlint lint --path main.swift --no-cache --enable-all-rules
Linting Swift files at paths main.swift
Linting 'main.swift' (1/1)
Done linting! Found 0 violations, 0 serious in 1 file.
```

### Environment

* SwiftLint version (run `swiftlint version` to be sure)? 0.43.1
* Installation method used (Homebrew, CocoaPods, building from source, etc)? portable
* Paste your configuration file:

```yml
```

* Are you using [nested configurations](https://github.com/realm/SwiftLint#nested-configurations)? No
  If so, paste their relative paths and respective contents.
* Which Xcode version are you using (check `xcodebuild -version`)? Xcode 12.4 Build version 12D4e
* Do you have a sample that shows the issue? Run `echo "[string here]" | swiftlint lint --no-cache --use-stdin --enable-all-rules`
  to quickly test if your example is really demonstrating the issue. If your example is more
  complex, you can use `swiftlint lint --path [file here] --no-cache --enable-all-rules`.

`main.swift`:

```swift
///
public class Test {
    ///
    public init(_ closure: () -> Void) {}

    deinit() {}
}

///
public let kTest1: Test = .init({})

///
public let kTest2: Test = .init {}
```

Running

```sh
sudo xcode-select --switch /Applications/Xcode-12.5.beta.3.app
```

and

```sh
./swiftlint lint --path main.swift --no-cache --enable-all-rules
```

prints

```
Linting Swift files at paths main.swift
Linting 'main.swift' (1/1)
/Users/revolt/1617901081/main.swift:10:27: warning: Trailing Closure Violation: Trailing closure syntax should be used whenever possible. (trailing_closure)
Done linting! Found 1 violation, 0 serious in 1 file.
```

**Comment by marcelofabri** [comment]
Looks like Swift 5.4 fixes a bug in SourceKit where `.init {}` wouldn't be returned as `source.lang.swift.expr.call` in the structure.

Unfortunately, I don't think there's much we can do here without changing the implementation a lot, which isn't worth it IMO given that this is a false negative (not a false positive) and it's been fixed in recent Swift versions.

**Comment by revolter** [comment]
I see. So the `SourceKit` was the cause. I think that we can close this issue then.
